File: examples/governance/operations_domain/lib/operations_domain/monitoring/commands/record_metric.rb
module OperationsDomain
  class Monitoring
    module Commands
      class RecordMetric
        include Hecks::Command
        emits "RecordedMetric"

        attr_reader :model_id
        attr_reader :deployment_id
        attr_reader :metric_name
        attr_reader :value
        attr_reader :threshold

        def initialize(
          model_id: nil,
          deployment_id: nil,
          metric_name: nil,
          value: nil,
          threshold: nil
        )
          @model_id = model_id
          @deployment_id = deployment_id
          @metric_name = metric_name
          @value = value
          @threshold = threshold
        end

        def call
          Monitoring.new(
            model_id: model_id,
            deployment_id: deployment_id,
            metric_name: metric_name,
            value: value,
            threshold: threshold,
            recorded_at: Time.now.to_s
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/operations_static_go/domain/record_metric.go
package domain

import (
	"time"
)

type RecordMetric struct {
	ModelId string `json:"model_id"`
	DeploymentId string `json:"deployment_id"`
	MetricName string `json:"metric_name"`
	Value float64 `json:"value"`
	Threshold float64 `json:"threshold"`
}

func (c RecordMetric) CommandName() string { return "RecordMetric" }

func (c RecordMetric) Execute(repo MonitoringRepository) (*Monitoring, *RecordedMetric, error) {
	agg := NewMonitoring(c.ModelId, c.DeploymentId, c.MetricName, c.Value, c.Threshold, time.Time{})
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := RecordedMetric{
		AggregateID: agg.ID,
		ModelId: c.ModelId,
		DeploymentId: c.DeploymentId,
		MetricName: c.MetricName,
		Value: c.Value,
		Threshold: c.Threshold,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
